### PR TITLE
[R4R] Incentive module simulations

### PR DESCRIPTION
**Pull request by denalimarsh**
This PR contains simulations for the incentive module. Base branch is set to `kd-usdx-incentives` for ease of review.

- [x] decoder
- [x] decoder tests
- [x] genesis
- [x] params
- [x] operations

MsgClaimReward operations have a low chance of meeting validity requirements. 

**1. With Seed = 52:**
```
 "incentive": {
  "claim_reward": {
   "ok": 106
  },
  "no-operation (no accounts currently have fulfillable claims)": {
   "failure": 304
  }
 },
```

**2. With Seed = 2:**
```
 "incentive": {
  "claim_reward": {
   "ok": 14
  },
  "no-operation (no accounts currently have fulfillable claims)": {
   "failure": 728
  }
 }
```

This PR is blocked until https://github.com/Kava-Labs/kava/pull/399 has been merged.

### CODE REVIEW

File: x/incentive/simulation/genesis.go [hunk 1]
```diff
@@ -0,0 +1,150 @@
...
+	authexported "github.com/cosmos/cosmos-sdk/x/auth/exported"
+	"github.com/cosmos/cosmos-sdk/x/simulation"
+
+	"github.com/kava-labs/kava/x/incentive/types"
+)
+
+var (
+	CollateralDenoms    = [3]string{"bnb", "xrp", "btc"}
+	RewardDenom         = "ukava"
+	MaxTotalAssetReward = sdk.NewInt(1000)
```
Comment (karzak) [inline x/incentive/simulation/genesis.go:20]: Increase this to 1000000000 - this is the total rewards per 24 or 48 hour period, so setting it this low would probably cause a lot of reward per second values to be 0. Related - let's set the minimum inflation in `kavadist` to 10%, so we can maximize the probability of having enough coins to payout rewards.

File: x/incentive/simulation/operations/msg.go [hunk 2]
```diff
@@ -0,0 +1,116 @@
...
+		r.Shuffle(len(openClaims), func(i, j int) {
+			openClaims[i], openClaims[j] = openClaims[j], openClaims[i]
+		})
+
+		// Randomly select a reward's collateral type from params
+		params := k.GetParams(ctx)
+		if len(params.Rewards) == 0 {
+			return noOpMsg, nil, fmt.Errorf("no rewards found in incentive module params")
+		}
+		reward := params.Rewards[r.Intn(len(params.Rewards))]
```
Comment (karzak) [inline x/incentive/simulation/operations/msg.go:60]: Don't choose a random reward denom, just try and claim whichever reward is found first. Shuffling the claims each time will provide sufficient randomness.

File: x/incentive/simulation/operations/msg.go [hunk 3]
```diff
@@ -0,0 +1,116 @@
...
+
+		// Load kavadist module account's current balance
+		kavadistMacc := sk.GetModuleAccount(ctx, kavadist.KavaDistMacc)
+		kavadistBalance := kavadistMacc.SpendableCoins(ctx.BlockTime())
+
+		// Find address that has a claim of the same reward denom, then confirm it's distributable
+		claimer, claim, found := findValidAccountClaimPair(accounts, openClaims, func(acc authexported.Account, claim types.Claim) bool {
+			if claim.Owner.Equals(acc.GetAddress()) { // Account must be claim owner
+				if claim.Denom == reward.Denom { // Claim's denom must match our reward's denom
```
Comment (karzak) [inline x/incentive/simulation/operations/msg.go:69]: see above, can skip this check

File: x/incentive/simulation/genesis.go [hunk 4]
```diff
@@ -0,0 +1,150 @@
...
+	for _, denom := range CollateralDenoms {
+		active := true
+		// total reward is in range (half max total reward, max total reward)
+		amount := simulation.RandIntBetween(r, int(MaxTotalAssetReward.Int64()/2), int(MaxTotalAssetReward.Int64()))
+		totalRewards := sdk.NewInt64Coin(RewardDenom, int64(amount))
+		// generate a random number of days between 1-2 to use for reward's times
+		numbDays := simulation.RandIntBetween(r, 1, 2) * 24
+		duration := time.Duration(time.Hour * time.Duration(numbDays))
+		timeLock := time.Duration(time.Hour * time.Duration(numbDays/2)) // half as long as duration
+		claimDuration := time.Hour * time.Duration(numbDays*2)           // twice as long as duration
```
Comment (karzak) [inline x/incentive/simulation/genesis.go:62]:
```suggestion
		numbHours := simulation.RandIntBetween(r, 6, 48)
		duration := time.Duration(time.Hour * time.Duration(numbHours))
		timeLock := time.Duration(time.Hour * time.Duration(numbHours/2)) // half as long as duration
		claimDuration := time.Hour * time.Duration(numbHours*2)           // twice as long as duration
```

Suggestion: shorten the minimum period length so there can be more reward claims per sim.

File: x/kavadist/simulation/genesis.go [hunk 5]
```diff
@@ -0,0 +1,84 @@
...
+	"github.com/cosmos/cosmos-sdk/x/simulation"
+
+	"github.com/kava-labs/kava/x/kavadist/types"
+)
+
+// SecondsPerYear is the number of seconds in a year
+const SecondsPerYear = 31536000
+
+// BaseAprPadding prevents the calculated SPR inflation rate from being 0.0
+const BaseAprPadding = "0.000000000100000000"
```
Comment (karzak) [inline x/kavadist/simulation/genesis.go:20]: To set minimum inflation to 10%, the BaseAprPadding would be `0.000000003022265980`

File: x/kavadist/simulation/genesis.go [hunk 6]
```diff
@@ -0,0 +1,84 @@
...
+		periods = append(periods, period)
+		periodStart = periodEnd
+	}
+	return periods
+}
+
+func genRandomInflation(r *rand.Rand) sdk.Dec {
+	// If sim.RandomDecAmount returns 0 (happens frequently by design), add BaseAprPadding
+	extraAprInflation := simulation.RandomDecAmount(r, sdk.MustNewDecFromStr("0.25"))
+	for extraAprInflation.Equal(sdk.ZeroDec()) {
```
Comment (karzak) [inline x/kavadist/simulation/genesis.go:64]:
```suggestion
	for extraAprInflation.LT(sdk.MustNewDecFromStr(BaseAprPadding)) {
```

File: x/incentive/simulation/genesis.go (hunk 1 above)
Comment (denalimarsh) [reply x/incentive/simulation/genesis.go]: https://github.com/Kava-Labs/kava/pull/439/commits/1b5f2889b5741ed550ff2f45a71f5a0e2d090b4e

File: x/kavadist/simulation/genesis.go (hunk 5 above)
Comment (denalimarsh) [reply x/kavadist/simulation/genesis.go]: https://github.com/Kava-Labs/kava/pull/439/commits/1b5f2889b5741ed550ff2f45a71f5a0e2d090b4e

File: x/incentive/simulation/operations/msg.go (hunk 2 above)
Comment (denalimarsh) [reply x/incentive/simulation/operations/msg.go]: https://github.com/Kava-Labs/kava/pull/439/commits/b47bb2f59a9bd19d17a0f6e6e0d4946dfb4b79aa

File: x/incentive/simulation/operations/msg.go (hunk 3 above)
Comment (denalimarsh) [reply x/incentive/simulation/operations/msg.go]: https://github.com/Kava-Labs/kava/pull/439/commits/b47bb2f59a9bd19d17a0f6e6e0d4946dfb4b79aa

File: x/incentive/simulation/genesis.go (hunk 4 above)
Comment (denalimarsh) [reply x/incentive/simulation/genesis.go]: https://github.com/Kava-Labs/kava/pull/439/commits/b47bb2f59a9bd19d17a0f6e6e0d4946dfb4b79aa

File: x/kavadist/simulation/genesis.go (hunk 6 above)
Comment (denalimarsh) [reply x/kavadist/simulation/genesis.go]: https://github.com/Kava-Labs/kava/pull/439/commits/b47bb2f59a9bd19d17a0f6e6e0d4946dfb4b79aa

**Comment by denalimarsh** [comment]
@karzak after the revisions, I get the following for `-Seed=2`:
```
"incentive": {
  "claim_reward": {
   "failure": 10,
   "ok": 34
  },
  "no-operation (no accounts currently have fulfillable claims)": {
   "failure": 698
  }
```
The failures are due to the kavadist module account having an insufficent balance to pay out the claim. I'm checking for this as part of `findValidAccountClaimPair()`, but there seems to be a mismatch. I've pushed a demo of the issue in  https://github.com/Kava-Labs/kava/pull/439/commits/bf9055c8461ff1de3da91d978360955ea7448c3b, running it logs relevant output i.e.
```
claim reward: 2859ukava
kavadist balance: 601716 ukava
Failed: {"codespace":"incentive","code":5,"message":"module account kavadist has insufficient balance to pay claim"}
```

Any idea why the module account appears to have enough tokens but the message still fails?

**Pull request merged**